src/python/pants/build_graph/target.py

```diff
@@ -868,6 +868,13 @@ def create_sources_field(self, sources, sources_rel_path, key_arg=None):
       # This is so that tests don't need to set up the subsystem when creating targets that
       # legitimately do not require sources.
       if (key_arg is None or key_arg == 'sources') and self.supports_default_sources():
+        deprecated_conditional(
+          lambda: True,
+          '1.10.0.dev0',
+          ('Specifying default sources on Targets is deprecated, and now takes a slow path. '
+           'Instead, class {} should have a TargetAdaptor installed specifying default sources to '
```

Comment:
I don't think there is a way for a plugin author to do this, since `TargetAdaptors` are (intentionally) not exposed as an extension point right now... see #3560 (mentioned in src/python/pants/init/engine_initializer.py) and #4535 about that.

I think that in order to do this, we'd need to actually expose `TargetAdaptor`... and it's not clear how to do that. A few options:

1. doing it as a static/cls method of `Target`
2. requiring that `Target` types that need to be constructed with a `TargetAdaptor` subclass other than the default need to install `def rules` with a particular shape
3. ?